src/bookmarks.rs

```diff
@@ -0,0 +1,95 @@
+use crate::config::Flattenable;
+use crate::error::JoshutoResult;
+use serde_derive::Deserialize;
+use serde_derive::Serialize;
+use std::collections::HashMap;
+use std::io::Write;
+use std::path::PathBuf;
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct RawBookmarkMapping {
+    #[serde(default)]
+    pub bookmarks: Vec<(char, String)>,
+}
+
+impl RawBookmarkMapping {
+    fn save(&self, path: &str) -> std::io::Result<()> {
+        let s = toml::to_string_pretty(self);
+        if let Ok(s) = s {
+            let mut file = std::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(path)
+                .unwrap();
+
+            file.write_all(&s.as_bytes()).unwrap();
+        }
+        Ok(())
+    }
+}
+
+impl Flattenable<BookmarkMapping> for RawBookmarkMapping {
+    fn flatten(self) -> BookmarkMapping {
+        let mut bookmarksmap = BookmarkMapping::new();
```

Comment:
This can be reduced to
```rust
fn flatten(self) -> BookmarkMapping {
    let map: HashMap<char, PathBuf> = self.bookmarks
        .drain(..)
        .map(|(c, s)| (c, PathBuf::from(s))
        .collect();
    BookmarkMapping { map }
}
```